cxx-qt-lib/src/qdate.rs

```diff
@@ -0,0 +1,194 @@
+// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use cxx::{type_id, ExternType};
+use std::mem::MaybeUninit;
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub struct QDate {
+    jd: i64,
+}
+
+extern "C" {
+    #[link_name = "cxxqt1$qdate$init"]
+    fn qdate_init(this: &mut MaybeUninit<QDate>, y: i32, m: i32, d: i32);
+}
+
+struct ParsedDate {
+    year: i32,
+    month: i32,
+    day: i32,
+}
+
+impl Default for QDate {
+    fn default() -> Self {
+        Self::new(0, 0, 0)
+    }
+}
+
+impl QDate {
+    pub fn new(y: i32, m: i32, d: i32) -> Self {
+        let mut s = MaybeUninit::<QDate>::uninit();
+
+        // Safety:
+        //
+        // Static checks on the C++ side ensure that QDate has the
+        // same binary footprint in C++ and Rust.
+        unsafe {
+            qdate_init(&mut s, y, m, d);
+            s.assume_init()
+        }
+    }
+
+    pub fn is_null(&self) -> bool {
+        !self.is_valid()
+    }
+
+    pub fn is_leap_year(mut y: i32) -> bool {
+        // No year 0 in Gregorian calendar, so -1, -5, -9 etc are leap years
+        if y < 1 {
+            y += 1;
+        }
+
+        (y % 4 == 0 && y % 100 != 0) || y % 400 == 0
+    }
+
+    pub fn is_valid(&self) -> bool {
+        (self.jd > -784350574879) && (self.jd < 784354017364)
+    }
+
+    pub fn is_valid_date(year: i32, month: i32, day: i32) -> bool {
+        // There is no year 0 in the Gregorian calendar.
+        year != 0
+            && day > 0
+            && month > 0
+            && month <= 12
+            && day
+                <= (if month == 2 {
+                    if QDate::is_leap_year(year) {
+                        29
+                    } else {
+                        28
+                    }
+                } else {
+                    QDate::days_in_usual_month(month)
+                })
+    }
+
+    pub fn year(&self) -> i32 {
```

Comment:
wonder if we should somehow call the QDate::year on the C++ side and use it's response? Then we could do the same with year, month, day, set_date. This would allow for removing the duplication of Qt C++ code.

Reply:
Done this is the latest commit now.